Serialization options for JsonOutput

jwagenleitner opened the pull request.

jwagenleitner commented:
This is work in progress (still working on adapting `StreamingJsonBuilder`) but am hoping to get some feedback on the new feature.   Main goal was to allow a way to make `JsonOutput` configurable while not changing the existing static methods and at the same time try to not duplicate too much code or provide too confusing a way to provide the options.  All feedback welcome.

Commit 2c0336b is best viewed by itself and by excluding whitespace by adding `-w` at the git diff command line or `w=1` in the GitHub querystring.

Trying to address the following JIRA tickets with this:

GROOVY-6699: JSON slurper ability to ignore propertie/field names (at least the serializing part)
GROOVY-6854: JsonOutput should be configurable with ISO 8601 UTC time zone instead of GMT
GROOVY-7682: JsonBuilder or JsonOutput incompatible with JodaTime (for now via a closure converter)
GROOVY-7780: JSONBuilder/JSONSlurper: Option to Only Include NonNull values
GROOVY-7858: Make JsonBuilder configurable to not write entries with null values

PascalSchumacher commented:
I'm not a groovy-json user, but this looks good as far as I can tell. :+1:

graemerocher commented:
+1 these changes would be awesome, and if we can include https://issues.apache.org/jira/browse/GROOVY-6975 even better ;-)

graemerocher commented:
@jwagenleitner Sweet! #433 seems another candidate to combine with this change. Thoughts?

graemerocher commented on subprojects/groovy-json/src/main/java/groovy/json/JsonOutput.java:
```diff
@@ -601,4 +280,966 @@ public String toString() {
+        private final boolean hasExcludedFieldNames;
+        private final boolean hasExcludedFieldTypes;
+
+        private Generator(Options options) {
```
Could we make this `protected` so subclasses could further customize as needed?

jwagenleitner replied:
Yes, that's a good idea, I'll make the constructor and `write*` methods protected.

To me it's starting to feel like the generator should be an interface (e.g., `JsonGenerator`) and having an abstract and default implementation that can be extended if desired.  But I don't want to go overboard on designing this so would be interested in your opinion and @paulk-asert .

graemerocher replied:
It does feel like that. I do think we should consider extensibility that way if users face an issue with customization it can easily be resolved. Another method I would like protected is `getObjectProperties(Object object)` since frameworks may need to customize that if they add additional properties to objects that are internal. We face this problem a lot in Grails.

graemerocher commented on subprojects/groovy-json/src/main/java/groovy/json/JsonOutput.java:
```diff
@@ -601,4 +280,966 @@ public String toString() {
+        /**
+         * Serializes object and writes it into specified buffer.
+         */
+        private void writeObject(String key, Object object, CharBuf buffer) {
```
Also if we do make the constructor `protected` making it possible to overload individual `write*` methods make sense?

paulk-asert commented on subprojects/groovy-json/src/main/java/groovy/json/JsonOutput.java:
```diff
@@ -492,7 +171,7 @@ private static void writeIterator(Iterator<?> iterator, CharBuf buffer) {
     public static String prettyPrint(String jsonPayload) {
         int indentSize = 0;
-        // Just a guess that the pretty view will take 20 percent more than original.
+        // Just a guess that the pretty view will take a 20 percent more than original.
```
why is 'a' added in above comment? Should it be 'about' or just a spurious typo?

jwagenleitner replied:
Thanks for pointing that out I'll remove the `a`.  I missed that when fixing up the merge conflict caused by commit 7e8211fa314526ce.

paulk-asert commented on subprojects/groovy-json/src/main/java/groovy/json/JsonOutput.java:
```diff
@@ -601,4 +280,966 @@ public String toString() {
+         * @return true if this converter can successfully convert values of
+         *      the given type to a JSON value
+         */
+        boolean handles(Class<?> type) {
```
Does this method and the next two need to be package private?

jwagenleitner replied:
I think my rationale here was that since it was a private nested class that is only used internally as a helper that package-private was the appropriate visibility.  But since the override methods are public it may make sense to be consistent and make these public too.

paulk-asert commented:
+1, looks great - some minor comments added inline - I think extensibility does need to be considered as per Graeme's comments but if you think that is going to cause further delay, we can always do that as a separate PR/commit

paulk-asert commented:
Might be worth adding in a test for extensibility, e.g.:
{code}
/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package groovy.json.custom

import groovy.json.JsonObject
import groovy.json.JsonOutput
import groovy.json.internal.CharBuf

/**
 * Tests extensibility of JsonOutput and helpers
 */
class CustomJsonOutputTest extends GroovyTestCase {

    void testGeneratorExcludesNullValues() {
        def generator = CustomJsonOutput.options()
                .excludeNulls()
                .makeLower()
                .createGenerator()

        def json = generator.toJson(new JsonObject(name: 'Test', properties: null))
        assert json == '{"name":"test"}'

        json = generator.toJson([field1: null, field2: "TEST"])
        assert json == '{"field2":"test"}'

        def ref = new JsonObject(name: 'Jason', properties: ['foo': 'bAr'])
        assert generator.toJson(ref) == '{"properties":{"foo":"bar"},"name":"jason"}'

        def jsonArray = ["FoO", ["bar":"tESt","name":"Jane"], "BaZ"]
        def jsonExpected = '["foo",{"bar":"test","name":"jane"},"baz"]'
        assert generator.toJson(jsonArray) == jsonExpected
    }

    static class CustomJsonOutput extends JsonOutput {
        static CustomOptions options() {
             new CustomOptions()
        }
    }

    static class CustomOptions extends JsonOutput.Options {
        boolean makeLower

        @Override
        CustomOptions excludeNulls() {
            (CustomOptions) super.excludeNulls()
        }

        @Override
        JsonOutput.Generator createGenerator() {
            new CustomGenerator(this)
        }

        public CustomOptions makeLower() {
            makeLower = true
            this
        }
    }

    static class CustomGenerator extends JsonOutput.Generator {
        CustomGenerator(CustomOptions options) {
            super(options)
        }

        @Override
        protected void writeObject(String key, Object object, CharBuf buffer) {
            super.writeObject(key, object instanceof CharSequence ? object.toString().toLowerCase() : object, buffer)
        }
    }
}
{code}
It won't compile/run until the necessary constructors/methods are made protected. I didn't carry the exercise over to Converter but that could be done too I guess.

jwagenleitner commented:
@graemerocher @paulk-asert thanks for all the feedback.  I think I have addressed all the points in the latest commit 16ef9c9.  I pulled out the `Generator` (and `Options` builder) from `JsonOutput` and made it its own type along with a default implementation.  I have also added an extensibility test as Paul suggested.

MarcusJL commented:
+1, This looks like a nice solution for me and would also cover #433.
For which release would that be planned?

jwagenleitner commented:
@MarcusJL thanks for taking a look at this, I appreciate it.  Unless there is other feedback in the next week or so I'll plan on merging this into master, so this would be for the 2.5 release.

MarcusJL commented:
@jwagenleitner thanks, that sounds great!
